File: medplum-java/medplum-model/src/main/java/com/fhir/FlagStatus.java
package com.fhir;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Locale;
import javax.annotation.Nonnull;

public final class FlagStatus {
  public static final FlagStatus ACTIVE = new FlagStatus(Value.ACTIVE, "ACTIVE");

  public static final FlagStatus INACTIVE = new FlagStatus(Value.INACTIVE, "INACTIVE");

  private final Value value;

  private final String string;

  FlagStatus(Value value, String string) {
    this.value = value;
    this.string = string;
  }

  public Value getEnumValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return this.string;
  }

  @Override
  public boolean equals(Object other) {
    return (this == other) 
      || (other instanceof FlagStatus && this.string.equals(((FlagStatus) other).string));
  }

  @Override
  public int hashCode() {
    return this.string.hashCode();
  }

  public <T> T accept(Visitor<T> visitor) {
    switch (value) {
      case ACTIVE:
        return visitor.visitACTIVE();
      case INACTIVE:
        return visitor.visitINACTIVE();
      case UNKNOWN:
      default:
        return visitor.visitUnknown(string);
    }
  }

  @JsonCreator(
      mode = JsonCreator.Mode.DELEGATING
  )
  public static FlagStatus valueOf(@Nonnull String value) {
    String upperCasedValue = value.toUpperCase(Locale.ROOT);
    switch (upperCasedValue) {
      case "ACTIVE":
        return ACTIVE;
      case "INACTIVE":
        return INACTIVE;
      default:
        return new FlagStatus(Value.UNKNOWN, upperCasedValue);
    }
  }

  public enum Value {
    ACTIVE,

    INACTIVE,

    UNKNOWN
  }

  public interface Visitor<T> {
    T visitACTIVE();

    T visitINACTIVE();

    T visitUnknown(String unknownType);
  }
}

File: medplum-typescript/src/model/fhir/FlagStatus.ts
export type FlagStatus =
    | FlagStatus.Active
    | FlagStatus.Inactive;

export const FlagStatus = {
    Active: "active" as FlagStatus.Active,
    Inactive: "inactive" as FlagStatus.Inactive,

    _visit: <Result>(value: FlagStatus, visitor: FlagStatus._Visitor<Result>): Result => {
        switch (value) {
            case FlagStatus.Active: return visitor.active();
            case FlagStatus.Inactive: return visitor.inactive();
            default: return visitor._unknown();
        }
    },
};

export declare namespace FlagStatus {
    export type Active = "active" & {
        __Active: void,
    };
    export type Inactive = "inactive" & {
        __Inactive: void,
    };

    export interface _Visitor<Result> {
        active: () => Result;
        inactive: () => Result;
        _unknown: () => Result;
    }
}
